lib/errors.js

```diff
@@ -0,0 +1,267 @@
+//   context to include in stack traces. This and the previous arg are typically
+//   select from the agent `sourceLines{Error,Span}{App,Library}Frames` config
+//   vars.
+//
+// Optional args:
+// - `args.callSiteLoc` - A `Error.captureStackTrace`d object with a stack to
+//   include as `error.log.stacktrace`.
+// - `args.traceContext` - The current TraceContext, if any.
+// - `args.trans` - The current transaction, if any.
+// - `args.errorContext` - An object to be included as `error.context`.
+// - `args.message` - A message string that will be included as `error.log.message`
+//   if `args.exception` is given. Ignored if `args.logMessage` is given.
+//
+// This always calls back with `cb(null, apmError)`, i.e. it doesn't fail.
+function createAPMError (args, cb) {
+  let numAsyncStepsRemaining = 0 // finish() will call cb() only when this is 0.
+
+  const error = {
+    id: crypto.randomBytes(16).toString('hex'),
+    timestamp: args.timestampUs
+  }
+  if (args.traceContext) {
+    error.parent_id = args.traceContext.traceparent.id
+    error.trace_id = args.traceContext.traceparent.traceId
+  }
+  if (args.trans) {
+    error.transaction_id = args.trans.id
+    error.transaction = {
+      type: args.trans.type,
+      sampled: args.trans.sampled
+    }
+  }
+  if (args.errorContext) {
+    error.context = args.errorContext
+  }
+
+  if (args.exception) {
+    // Handle an exception, i.e. `captureError(<an Error instance>, ...)`.
+    const err = args.exception
+    const errMsg = String(err.message)
+    error.exception = {
+      message: errMsg,
+      type: String(err.name),
+      handled: args.handled
+    }
+
+    if ('code' in err) {
+      error.exception.code = String(err.code)
+    } else {
+      // To provide better grouping of mysql errors that happens after the async
+      // boundery, we modify to exception type to include the custom mysql error
+      // type (e.g. ER_PARSE_ERROR)
+      var match = errMsg.match(MYSQL_ERROR_MSG_RE)
+      if (match) {
+        error.exception.code = match[1]
+      }
+    }
+
+    // Optional add an alternative error message as well as the exception message.
+    if (args.message && typeof args.message === 'string') {
+      error.log = { message: args.message }
+    }
+
+    if (args.shouldCaptureAttributes) {
+      const attrs = attributesFromErr(err)
+      if (attrs) {
+        error.exception.attributes = attrs
+      }
+    }
+
+    numAsyncStepsRemaining++
+    gatherStackTrace(
+      args.log,
+      args.exception,
+      args.sourceLinesAppFrames,
+      args.sourceLinesLibraryFrames,
+      null, // filterCallSite
+      function (_err, stacktrace) {
+        // _err from gatherStackTrace is always null.
+
+        const culprit = culpritFromStacktrace(stacktrace)
+        if (culprit) {
+          error.culprit = culprit
+        }
+        const moduleName = _moduleNameFromFrames(stacktrace)
+        if (moduleName) {
```

Comment:
REVIEW-NOTE: This fixes a bug added in #1920 where the variable is `const moduleName`, but the if-block is `if (module) {`  -- write var name.